File: bench/src/matmul/matmul.py
#!/usr/bin/env python3
# matmul.py — CPU variant, numpy (bench/SPEC.md)
import sys
import numpy as np
n = int(sys.argv[1]) if len(sys.argv) > 1 else 512
k = np.arange(n*n, dtype=np.float64)
a = (1.0 + 0.000001*k).reshape(n, n)
b = (1.0 - 0.000001*k).reshape(n, n)
c = a @ b
print(f"c[0,0]: {c[0,0]:.6f}")
print(f"c[N-1,N-1]: {c[n-1,n-1]:.6f}")
print(f"checksum: {c.sum():.2f}")

File: bench/src/matmul/matmul.js
// matmul.js — CPU variant (bench/SPEC.md)
const n = parseInt(process.argv[2] || "512");
const a = new Float64Array(n*n), b = new Float64Array(n*n), c = new Float64Array(n*n);
for(let k=0;k<n*n;k++){ a[k] = 1.0 + 0.000001*k; b[k] = 1.0 - 0.000001*k; }
for(let i=0;i<n;i++) for(let k=0;k<n;k++){ const aik = a[i*n+k]; for(let j=0;j<n;j++) c[i*n+j] += aik*b[k*n+j]; }
let sum = 0; for(let k=0;k<n*n;k++) sum += c[k];
console.log(`c[0,0]: ${c[0].toFixed(6)}`);
console.log(`c[N-1,N-1]: ${c[n*n-1].toFixed(6)}`);
console.log(`checksum: ${sum.toFixed(2)}`);
